### PR TITLE
Curtain sign work

### PR DESCRIPTION
**Pull request by delaneysmith22**
This is the first PR related to curtain sign classification. The proposed workflow is based on a combination of view classification and time series classification:
1. Deploy the view classifier to generate a pleural prediction probability time series dataset
2. Train a 1D time series classification model to distinguish the curtain sign artifact, given its (hopefully unique) oscillatory signature in this reduced feature space

The code is written to support both binary (curtain sign vs. else) and multiclass (curtain sign vs. blind sign vs. pathology) classification. Note that the blind sign refers to clips with no curtain sign, but also no pathology. The new files that have been included in the PR are described below.
- `curtain_cfg.yml`: config file specific to curtain sign classification
- ` build_curtain_dataset.py`: script to build a pleural prediction probability time series dataset from a raw database query
- `train_curtain.py`: script to train a time series classification model. Includes the option to train a simple k-nearest neighbors with dynamic time warping model to use a baseline for performance as well as a 1D custom CNN model.
- `curtain_models.py`: script to contain all DL classifiers to be considered during model development. Currently contains a custom 1D CNN model which an arbitrary structure (to be updated in future PRs to facilitate hyperparameter searches to find the best performing architecture for our classification task). 
- `predict_curtain.py`: prediction script specific to curtain sign classification
- `curtain_utils.py`: script containing an assortment of utility functions relating to curtain sign classification
- `curtain_visualization.py`: visualization script specific to curtain sign classification

### CODE REVIEW

File: src/data/build_curtain_dataset.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+    :param random_state: An integer for setting the random seed for partitioning.
+
+    :return df: Dataframe with split labels (if partition = True) or without split labels (if partition = False)
+    '''
+
+    # Load database query as DataFrame
+    df = pd.read_csv(database_query)
+
+    # Sample pleural clips only
+    df = df.loc[df['view'] == 'pleural']
```
Comment (bvanber) [inline src/data/build_curtain_dataset.py:36]: Is there a reason we are filtering by view via code instead of only querying for the view we want?

File: src/data/build_curtain_dataset.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+    df1.drop(df1.loc[df1['reviewed_class'] == 0].index, axis=0, inplace=True)
+    df2.drop(df2.loc[(df2['reviewed_class'] == 0) | (df2['reviewed_class'] == 1)].index, axis=0, inplace=True)
+
+    # Add mislabelled clips to their correct dfs
+    df0 = pd.concat([df0, extra0])
+    df1 = pd.concat([df1, extra1])
+
+    # Impose correct labels
+    df0['curtain_sign'] = 1
+    df0['pleural_effusion'] = np.nan
```
Comment (bvanber) [inline src/data/build_curtain_dataset.py:303]: Can we add dome documentation as to why we are saving the nan values here please

File: src/data/build_curtain_dataset.py [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+    npzs_dir = curtain_cfg['PATHS']['NPZS']
+    np.savez_compressed(os.path.join(npzs_dir, 'gold_dataset_' + datetime.datetime.now().strftime("%Y%m%d-%H%M%S")),
+                        dataset=new_dataset)
+
+    return new_df
+
+if __name__ == "__main__":
+
+    # Create dataset
+    frame_preds = 'data/curtain_sign/view_predictions/frame_predictions20221103-143454.csv'
```
Comment (bvanber) [inline src/data/build_curtain_dataset.py:361]: I dont see this being used, shoud either be removed or turned into a cfg field

File: src/data/query_to_df.py [hunk 4]
```diff
@@ -45,10 +40,13 @@ def create_frame_dataframe(database_query):
 
     df['s3_path'] = df.apply(lambda row: row.s3_path, axis=1)
 
+    df.drop_duplicates(subset=['s3_path'], keep='first', inplace=True)
```
Comment (bvanber) [inline src/data/query_to_df.py:43]: Why do we need this? Have we observed duplicates?

File: src/data/build_dataset.py [hunk 5]
```diff
@@ -1,80 +1,89 @@
...
+                                    'Class Name': cfg['DATA']['CLASSES'][row['class']],
+                                    'filename': row['s3_path'].split('/')[-1][:-4]})
+            clip_dfs.append(clip_df)
+    all_clips_df = pd.concat(clip_dfs, axis=0, ignore_index=True)
+    if fold is not None:
+        all_clips_df.to_csv('data/fold_{}_frames_table.csv'.format(fold), index=False)
+
+    else:
+        #og_frames = pd.read_csv('data/non_curtain_normal_frames_table.csv')
+        #all_clips_df = pd.concat([all_clips_df,og_frames])
```
Comment (bvanberl) [inline src/data/build_dataset.py:71]: Are we able to remove the above comments?

File: src/data/curtain_utils.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    '''
+    Copies previously generated pngs with specific filenames (stored in df) to a target directory.
+
+    :param df: Dataframe with the png filenames to be copied.
+    :param origin: Path to directory containing .pngs.
+    :param target: Directory where .pngs will be stored.
+    '''
+    for file_name in list(df['filename']):
+        if file_name + '.png' in os.listdir(origin):
+            shutil.copy(os.path.join(origin, file_name + '.png'), os.path.join(target, file_name + '.png'))
```
Comment (bvanberl) [inline src/data/curtain_utils.py:18]: Wondering if we should print something out to alert the user in case of files that don't exist. It might be useful to know if 600 images weren't copied because they don't exist.

File: src/data/curtain_utils.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+def copy_pngs(df, origin, target):
+    '''
+    Copies previously generated pngs with specific filenames (stored in df) to a target directory.
+
+    :param df: Dataframe with the png filenames to be copied.
+    :param origin: Path to directory containing .pngs.
+    :param target: Directory where .pngs will be stored.
+    '''
+    for file_name in list(df['filename']):
+        if file_name + '.png' in os.listdir(origin):
```
Comment (bvanberl) [inline src/data/curtain_utils.py:17]: You could also use `if os.path.exists(file_name + '.png')` instead of explicitly looping through the files in the directory.

File: src/data/database_pull.py [hunk 8]
```diff
@@ -6,7 +6,7 @@
 
 cfg = yaml.full_load(open(os.getcwd() + "/config.yml", 'r'))
 
-def data_pull(fold=None):
+def data_pull(cfg, fold=None):
     '''
     Pull raw clips from AWS database using a generated query in csv format
     '''
```
Comment (bvanberl) [inline src/data/database_pull.py:12]: Missing argument docs

File: src/data/query_to_df.py [hunk 9]
```diff
@@ -26,13 +26,8 @@ def create_frame_dataframe(database_query):
...
 
-    # Create filename
-    df['filename'] = df['exam_id'] + "_" + df['patient_id'] + "_VID" + df['vid_id']
-
-    for i, name in enumerate(df['filename']):
-        if 'nan' in name:
-            df['filename'][i] = "vid{}_vid{}_VID{}".format(i, i, i)
-            print(i, name, df['filename'][i])
+    # Create filename from s3_path
+    df['filename'] = df.apply(lambda row: row['s3_path'].split('/')[-1][:-4],axis=1)
```
Comment (bvanberl) [inline src/data/query_to_df.py:30]: Extracting a filename from an S3 path is certainly a useful thing. Wondering if it should be a separate function that we import as a utility.

File: src/models/curtain_models.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    model.add(BatchNormalization())
+    model.add(Dropout(0.5))
+
+    # Convolutional block 2
+    model.add(Conv1D(256, 2, strides=1, activation='relu'))
+    model.add(Conv1D(256, 2, strides=1, activation='relu'))
+    model.add(Conv1D(256, 2, strides=1, activation='relu'))
+
+    model.add(BatchNormalization())
+    model.add(Dropout(0.5))
```
Comment (bvanberl) [inline src/models/curtain_models.py:57]: I think it's more common (certainly in our work) to place dropout *after* global pooling, *I think* since information loss is already occurring in global pooling. But I don't think there's a research consensus as to which option is better, if any. It would be interesting to investigate whether this has an effect.

File: src/predict.py [hunk 11]
```diff
@@ -290,9 +300,13 @@ def compute_frame_predictions(cfg, dataset_files_path, class_thresh=0.5, calcula
     pred_probs_df = pd.DataFrame(pred_probs, columns=['Pleural View Probability'])
     pred_probs_df.insert(0, 'Frame Path', files_df['Frame Path'])
     pred_probs_df.insert(1, 'Class', files_df['Class'])
+    pred_probs_df.insert(2, 'filename', files_df['filename'])
```
Comment (bvanberl) [inline src/predict.py:303]: You might be able to do the above using concatenation:
`pred_probs_df = pd.concat([pred_probs_df, files_df['Frame Path'], ...], axis=1)`

File: src/visualization/visualization.py [hunk 12]
```diff
@@ -33,21 +35,27 @@ def plot_to_tensor():
     return image_tensor
 
 
-def get_roc_data(path_to_preds, base_fpr=np.linspace(0, 1, 1001)):
+def get_roc_data(path_to_preds=None, labels=None, preds=None, base_fpr=np.linspace(0, 1, 1001)):
     '''
     Computes the true positive rate for plotting an ROC curve and the curves corresponding AUC.
     :param path_to_preds: path to frame-level prediction probabilities.
     :param base_fpr: fpr-vector that will be used for plotting.
```
Comment (bvanberl) [inline src/visualization/visualization.py:42]: Missing some parameter docs

**Review by bvanberl — APPROVED** [review]

File: src/models/curtain_models.py (hunk 10 above)
Comment (bvanber) [reply src/models/curtain_models.py:57]: Also I think the dropout rate should be in the config if we want to run some experiments

Comment (sdhossain) [reply src/models/curtain_models.py:57]: to add on the configurability, would the number of kernels/filters also be something of interest. (Also I am not sure if this is relevant to the PR exactly (apologies if its not), but would it be worth it to use tf.keras functional api or some other way of modularizing the blocks separately, perhaps even adding residual blocks).

File: src/train_curtain.py [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+
+    thresholds = 1.0 / cfg['TRAIN']['N_CLASSES']
+
+    # Define performance metrics
+    if binary:
+        metrics = ['accuracy',
+                   AUC(name='auc'),
+                   Precision(name='precision', thresholds=thresholds),
+                   Recall(name='recall', thresholds=thresholds)]
+    else:
```
Comment (sdhossain) [inline src/train_curtain.py:277]: I remember having some bugs with accuracy a while back, but since binary and categorical are separate, I ended up using binary_accuracy from https://www.tensorflow.org/api_docs/python/tf/keras/metrics/BinaryAccuracy.

File: src/models/curtain_models.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+    :param model_name: Name of model (str) to be returned.
+    :return model_def: TF model definition function
+    '''
+
+    if model_name == 'custom_1DCNN':
+        model_def = build_custom_1DCNN
+    else:
+        raise Exception("Invalid entry in TRAIN > MODEL_DEF field of config.yml.")
+    return model_def
+
```
Comment (sdhossain) [inline src/models/curtain_models.py:23]: this might be my personal preference, but I think a dictionary would look cleaner/serve the same purpose. (Especially with more model_names)

**Review by sdhossain — COMMENTED** [review]
Overall seems good to me! (I left a few comments)

File: src/data/build_curtain_dataset.py (hunk 1 above)
Comment (delaneysmith22) [reply src/data/build_curtain_dataset.py:36]: I've just added this as a precaution, in case parenchymal clips somehow were included in the query.

File: src/data/query_to_df.py (hunk 4 above)
Comment (delaneysmith22) [reply src/data/query_to_df.py:43]: There was a bug in the database pipeline that I noticed when I started this project because there were duplicates. I think this has been resolved now, so we shouldn't need this in the future, but it doesn't necessarily hurt to keep it - thoughts? @bvanber

File: src/data/build_curtain_dataset.py [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+        quality = row['quality']
+        if type(quality) == str:
+            for marker in quality_markers:
+                if marker in quality:
+                    df.loc[idx, marker] = 1
+
+        # Add class labels
+        curtain = row['curtain_sign'] == 1
+        effusion = type(row['pleural_effusion']) == str  # only a string if pe is present
+        consolidation = type(row['consolidation']) == str
```
[the remark this answers is not in the document]
Comment (delaneysmith22) [reply src/data/build_curtain_dataset.py:79]: If effusion and consolidation are not strings then they are nan, yes

File: src/data/build_curtain_dataset.py [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+
+    df['s3_path'] = df.apply(lambda row: row.s3_path, axis=1)
+
+    # Add columns for key quality markers, to simplify stratifying the dataset later
+    quality_markers = ['diaphragm_cannot_be_identified', 'significant_probe_movement',
+                       'poor_composition', 'borderline_composition', 'abdominal_focus']
+    df[quality_markers] = 0
+    df['small_basal_consolidation_only'] = 0
+
+    # Assign curtain sign class labels
```
[the remark this answers is not in the document]
Comment (delaneysmith22) [reply src/data/build_curtain_dataset.py:66]: Maybe I'm misunderstanding, but I don't think this will work. I want to set row[_marker_] equal to 1 based on whether _marker_ is a substring of row['quality'] (_marker_ doesn't need to equal row['quality'], it just has to be contained in it). Please correct me if I'm wrong or if there's a better way to do this @bvanberl.

File: src/models/curtain_models.py (hunk 10 above)
Comment (delaneysmith22) [reply src/models/curtain_models.py:57]: This model is currently just a placeholder to ensure that train_curtain.py runs. In the next PR, which relates to running hyperparameter sweeps in wandb, it's architecture and these parameters (dropout, kernels, filters) will be completely configurable.  I've added a line in the doc string to address this!

File: src/data/query_to_df.py (hunk 4 above)
Comment (bvanberl) [reply src/data/query_to_df.py:43]: Actually I completely agree that it's nice to have this secondary check. Perhaps you could print out the number of records before and after deduplication so that we are alerted in case it happens again.

File: src/data/build_curtain_dataset.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (bvanberl) [reply src/data/build_curtain_dataset.py:66]: Oh I see what you are trying to do. Pandas does provide some support for substring checking. I wonder if something like below could help, operating on all rows at once:
```
for marker in quality_markers:
    d[marker] = d['quality'].str.contains(marker).astype(int)
```

File: src/data/build_curtain_dataset.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (bvanberl) [reply src/data/build_curtain_dataset.py:79]: Oh I see. And they are nan if they are blanks I imagine.

A safer way to check for this (again with no row loop) might be `effusion = df['pleural_effusion'].notna().astype(int)`

File: src/data/build_curtain_dataset.py [hunk 17]
```diff
@@ -0,0 +1,367 @@
...
+            for marker in quality_markers:
+                if marker in quality:
+                    df.loc[idx, marker] = 1
+
+        # Add class labels
+        curtain = row['curtain_sign'] == 1
+        effusion = type(row['pleural_effusion']) == str  # only a string if pe is present
+        consolidation = type(row['consolidation']) == str
+        if curtain and not effusion and not consolidation:
+            df.loc[idx, 'curtain_class'] = 0
```
Comment (bvanberl) [inline src/data/build_curtain_dataset.py:81]: Putting it all together, outside the loop you could probably do this with
```
df['curtain_class'] = ((df['curtain_sign'] == 1) & df['pleural_effusion'].isna() & df['consolidation'].isna()).astype(int)
```

**Review by bvanberl — APPROVED** [review]
